classloader/src/main/java/io/smallrye/common/classloader/ClassDefiner.java

```diff
@@ -0,0 +1,41 @@
+package io.smallrye.common.classloader;
+
+import java.lang.invoke.MethodHandles;
+import java.lang.reflect.Field;
+import java.security.AccessController;
+import java.security.PrivilegedAction;
+
+import sun.misc.Unsafe;
+
+public class ClassDefiner {
+    private static final Unsafe unsafe;
+
+    static {
+        unsafe = AccessController.doPrivileged(new PrivilegedAction<Unsafe>() {
+            public Unsafe run() {
+                try {
+                    final Field field = Unsafe.class.getDeclaredField("theUnsafe");
+                    field.setAccessible(true);
+                    return (Unsafe) field.get(null);
+                } catch (IllegalAccessException e) {
+                    throw new IllegalAccessError(e.getMessage());
+                } catch (NoSuchFieldException e) {
+                    throw new NoSuchFieldError(e.getMessage());
+                }
+            }
+        });
+    }
+
+    public static Class<?> defineClass(MethodHandles.Lookup lookup, Class<?> parent, String className, byte[] classBytes) {
+        // The className for Unsafe.defineClass must match the one in classBytes, so we can try to verify the package
```

Comment:
The possibility exists that this code could be consumed from WildFly or other environments that require security manager support.  Since this operation is implicitly privileged, we need to add a permission check here.

Reply:
something like `new RuntimePermission("defineClass")` or do you prefer another name?

Reply:
Don't use RuntimePermission because it has a specific meaning in the JVM already.  I'd just create a `DefineClassPermission `and have it extend `Permission` with no action string supported.  That'd be the simplest & safest solution.

Reply:
Ok!